File: Linear Structures/Singly-Linked-Lists.java
//Java

//Node Class
public class Node {
	Node next;
	int value;
	public Node(int value) {
		this.value = value;
	}
}

//LinkedList Class
public class LinkedList {
	Node head;
	
	public void append(int value) {
		if (head == null) {
			head = new Node(value);
			return;
		}
		Node current = head;
		while (current.next != null) {
			current = current.next;
		}
		current.next = new Node(value);
	}
	
	public void prepend(int value) {
		Node newHead = new Node(value);
		newHead.next = head;
		head = newHead;
	}
	
	public void deleteFirst(int value) {
		if (head == null) {
			return;
		}
		if (head.value == value) {
			head = head.next;
			return;
		}
		Node current = head;
		while (current.next != null) {
			if (current.next.value == value) {
				current.next = current.next.next;
				return;
			}
			current = current.next;
		}
	}
	
	public void deleteAll(int value) {
		if (head == null) {
			return;
		}
		Node current = head;
		while (current.next != null) {
			if (current.next.value == value) {
				current.next = current.next.next;
			}
			else{
				current = current.next;
			}
		}
		if (head.value == value) {
			head = head.next;
		}	
	}
	
	public int getLength() {
		if (head == null) {
			return 0;
		}
		int count = 0;
		Node current = head;
		while(current != null) {
			count++;
			current = current.next;
		}
		return count;
	}
	
	public int[] getValues() {
		Node current = head;
		int[] rValues = new int[getLength()];
		
		if (head == null) {
			return rValues;
		}
		int i = 0;
		while (current != null) {
			rValues[i] = current.value;
			i++;
			current = current.next;
		}
		return rValues;
	}
	
	public void reverse() {
		Node prev = null;
		Node current = head;
		Node next;
		
		while(current != null) {
			next = current.next;
			current.next = prev;
			prev = current;
			current = next;
		}
		head = prev;
	}
	
	public void insert(int value, int position) {
		int counter = 0;
		Node current = head;
		if (counter == position) {
			prepend(value);
			return;
		}
		else {
			while(counter + 1 < position && current.next != null) {
				current = current.next;
				counter++;
			}
		}
		if (current.next == null) {
			return;
		}
		Node next = current.next;
		Node inserted = new Node(value);
		current.next = inserted;
		inserted.next = next;
	}
}

//Main File
import java.util.ArrayList;
public class main {
	public static void main(String[] args) {
		LinkedList l = new LinkedList();
		l.append(0);
		l.append(1);
		l.append(2);
		l.append(3);
		l.append(4);
		l.append(0);
		int[] lList;
		// 0 -> 1 -> 2 -> 3 -> 4 -> 0 -> NULL
		System.out.println(l.getLength());
		printLinkedList(l);
		l.deleteAll(0);
		//1 -> 2 -> 3 -> 4 -> NULL
		System.out.println();
		System.out.println(l.getLength());
		printLinkedList(l);
		System.out.println();
		l.prepend(1);
		//1 -> 1 -> 2 -> 3 -> 4 -> NULL
		printLinkedList(l);
		l.deleteAll(1);
		//2 -> 3 -> 4 -> NULL
		System.out.println();
		printLinkedList(l);
		l.prepend(0);
		System.out.println();
		printLinkedList(l);
		//0 -> 2 -> 3 -> 4 -> NULL
		l.reverse();
		System.out.println();
		printLinkedList(l);
		//4 -> 3 -> 2 -> 0 -> NULL
		l.reverse();
		System.out.println();
		printLinkedList(l);
		//0 -> 2 -> 3 -> 4 -> NULL
		l.append(0);
		//0 -> 2 -> 3 -> 4 -> 0 -> NULL
		System.out.println();
		printLinkedList(l);
		l.deleteFirst(0);
		// 2 -> 3 -> 4 -> 0 -> NULL
		System.out.println();
		printLinkedList(l);
		l.insert(1, 1);
		//2 -> 1 -> 3 -> 4 -> 0 -> NULL
		System.out.println();
		printLinkedList(l);
		l.insert(9, 0);
		//9 -> 2 -> 1 -> 3 -> 4 -> 0 -> NULL
		System.out.println();
		printLinkedList(l);
	}
	public static void printLinkedList(LinkedList k) {
		int[] lList = k.getValues();
		for (int i = 0; i < lList.length; i++) {
			System.out.print(lList[i] + ", ");
		}
	}
}

File: Linear Structures/Singly-Linked-Lists.cpp
// Node.hpp
struct Node {
    public:
        int value;
        Node* next;
        Node(int val);
};

//Node.cpp
#include "Node.hpp"
#include <iostream>
Node::Node(int val){
    value = val;
    next = NULL;
}

//LinkedList.hpp
#include "Node.cpp"
#include <iostream>
#include <vector>
using namespace std;
class LinkedList{
    public:
        Node* head;
        LinkedList();
        void append(int value);
        void prepend(int value);
        void deleteFirst(int value);
        void deleteAll(int value);
        int getLength();
        vector<int> getValues();
        void reverse();
        void insert(int value, int position);
};

//LinkedList.cpp
#include "LinkedList.hpp"
#include <iostream>
using namespace std;
LinkedList::LinkedList(){
    head = NULL;
}

void LinkedList::append(int value){
    if (head == NULL){
        head = new Node(value);
        return;
    }
    Node* current = head;
    while (current->next != NULL){
        current = current->next;
    }
    current->next = new Node(value);
}

void LinkedList::prepend(int value){
    Node* newHead = new Node(value);
    newHead->next = head;
    head = newHead;
}

void LinkedList::deleteFirst(int value){
    if (head == NULL){
        return;
    }
    if (head->value == value){
        head = head->next;
        return;
    }
    Node* current = head;
    while (current->next != NULL){
        if (current->next->value == value){
            current->next = current->next->next;
            return;
        }
        current = current->next;
    }
}

void LinkedList::deleteAll(int value){
    if (head == NULL){
        return;
    }
    Node* current = head;
    while (current->next != NULL){
        if (current->next->value == value){
            current->next = current->next->next;
        }
        else{
            current = current->next;
        }
    }
    if (head->value == value){
        head = head->next;
    }
}

int LinkedList::getLength(){
    if (head == NULL){
        return 0;
    }
    int count = 0;
    Node* current = head;
    while (current != NULL){
        count++;
        current = current->next;
    }
    return count;
}

std::vector<int> LinkedList::getValues(){
    Node* current = head;
    std::vector<int> rValues;

    if (head == NULL){
        return rValues;
    }
    while (current != NULL){
        rValues.push_back(current->value);
        current = current->next;
    }
    return rValues;
}

void LinkedList::reverse(){
    Node* prev = NULL;
    Node* current = head;
    Node* next;

    while (current != NULL){
        next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }
    head = prev;
}

void LinkedList::insert(int value, int position){
    int counter = 0;
    Node* current = head;
    if (counter == position){
        prepend(value);
        return;
    }
    else{
        while (counter + 1 < position && current->next != NULL){
            current = current->next;
            counter++;
        }
    }
    if (current->next == NULL){
        return;
    }
    Node* next = current->next;
    Node* inserted = new Node(value);
    current->next = inserted;
    inserted->next = next;
}

// main.cpp
#include <iostream>
#include "LinkedList.cpp"
using namespace std;

void printList(LinkedList list){
    std::vector<int> ll = list.getValues();
    for(int i = 0; i < ll.size(); i++){
        std::cout << ll[i] << " -> ";
    }
    std::cout << "\n";

}

int main(){
    LinkedList l;
    l.append(0);
    l.append(1);
    l.append(2);
    l.append(3);
    l.append(4);
    l.append(0);
    printList(l);
    l.deleteAll(0);
    printList(l);
    l.prepend(1);
    printList(l);
    l.deleteAll(1);
    printList(l);
    l.prepend(0);
    printList(l);
    l.reverse();
    printList(l);
    l.reverse();
    printList(l);
    l.append(0);
    printList(l);
    l.deleteFirst(0);
    printList(l);
    l.insert(1, 1);
    printList(l);
    l.insert(9, 0);
    printList(l);
}
